godot-core/src/engine/gfile.rs

```diff
@@ -0,0 +1,823 @@
+    ) -> IoResult<Self> {
+        let path: GString = path.into();
+        let fa = FileAccess::open_compressed_ex(path.clone(), flags)
+            .compression_mode(compression_mode)
+            .done()
+            .ok_or_else(|| {
+                IoError::new(
+                    ErrorKind::Other,
+                    format!(
+                        "can't open file {} in mode {:?}; GodotError: {:?}",
+                        &path,
+                        flags,
+                        FileAccess::get_open_error()
+                    ),
+                )
+            })?;
+
+        Ok(Self::from_inner(fa))
+    }
+
+    /// Open a file encrypted by byte key
+    ///
+    /// Opens a file encrypted by 32-byte long [`PackedByteArray`] located at
+    /// `path`, creating new [`GFile`] object.
+    /// For [`ModeFlags`] description check the [`GFile`] documentation.
+    pub fn open_encrypted(
+        path: impl Into<GString>,
+        flags: ModeFlags,
+        key: PackedByteArray,
+    ) -> IoResult<Self> {
+        let path: GString = path.into();
+        let fa = FileAccess::open_encrypted(path.clone(), flags, key).ok_or_else(|| {
+            IoError::new(
+                ErrorKind::Other,
+                format!(
+                    "can't open file {} in mode {:?}; GodotError: {:?}",
+                    &path,
+                    flags,
+                    FileAccess::get_open_error()
+                ),
+            )
+        })?;
+
+        Ok(Self::from_inner(fa))
+    }
+
+    /// Open a file encrypted by password
+    ///
+    /// Opens a file encrypted by a `pass` located at `path`, creating new
+    /// [`GFile`] object. For [`ModeFlags`] description check the [`GFile`]
+    /// documentation.
+    pub fn open_encrypted_with_pass(
+        path: impl Into<GString>,
+        flags: ModeFlags,
+        pass: GString,
+    ) -> IoResult<Self> {
+        let path: GString = path.into();
+        let fa =
+            FileAccess::open_encrypted_with_pass(path.clone(), flags, pass).ok_or_else(|| {
+                IoError::new(
+                    ErrorKind::Other,
+                    format!(
+                        "can't open file {} in mode {:?}; GodotError: {:?}",
+                        &path,
+                        flags,
+                        FileAccess::get_open_error()
+                    ),
+                )
+            })?;
+        Ok(Self::from_inner(fa))
+    }
+
+    /// Creates new [`GFile`] from provided [`Gd`]<[`FileAccess`]>.
+    ///
+    /// For this method to work, the provided `file_access` must be unique - no
+    /// other symbol can reference it. Its state is retained: [`ModeFlags`] with
+    /// which it was open and current internal cursor position.
+    pub fn from_unique(file_access: Gd<FileAccess>) -> Result<Self, NotUniqueError> {
+        let file_access = NotUniqueError::check(file_access)?;
+        Ok(Self::from_inner(file_access))
+    }
```

Comment:
I just noticed, we should also check `file_access.is_open()` and return `Err` otherwise.
Since this required re-modeling the type, could you add a `TODO` for now?